torch/csrc/jit/codegen/cuda/lower_unroll.cpp

```diff
@@ -32,6 +32,221 @@ kir::ForLoop* cloneLoopNest(const kir::ForLoop* for_loop, bool unroll = false) {
+
+  return namedScalar;
+}
+
+kir::ForLoop* handleMisalignedVectorization(
+    std::vector<kir::ForLoop*> loop_structure,
+    const kir::ForLoop* for_loop) {
+  // body -> allocate, read, compute, write for-loops
+  kir::IrBuilder ir_builder(GpuLower::current()->kernel());
+  kir::Int* zero = ir_builder.create<kir::Int>(0);
+
+  // create new base For-Loop
+  const auto new_loop = ir_builder.create<kir::ForLoop>(
+      for_loop->index(), for_loop->extent(), for_loop->iter_domain());
+
+  // Find child For-Loops and add remaining expressions to base For-Loop
+  auto child_loops = parseVectorizedForLoop(for_loop, new_loop);
+
+  // Find the first vectorize set - either read or write
+  // Add child For-Loop to loop_structure
+  // Enable vectorize flag in child For-Loop
+  auto vec_expr = findVectorizedSet(loop_structure, child_loops);
+  TORCH_INTERNAL_ASSERT(vec_expr != nullptr);
+
+  // out_tv is the TensorView with the misaligned vec iterDomain
+  auto out_tv = vec_expr->outputs().front()->as<kir::TensorView>();
+  auto in_tv = vec_expr->inputs().front()->as<kir::TensorView>();
+  kir::TensorView* vec_tv =
+      (out_tv->memoryType() == MemoryType::Local) ? out_tv : in_tv;
+
+  const auto pred = PredicateCompute::getInlinePredicate(
+      vec_expr, loop_structure, nullptr, false, true);
+
+  kir::IfThenElse* pred_ite = ir_builder.create<kir::IfThenElse>(pred);
+  new_loop->body().push_back(pred_ite);
+
+  // Generate vectorize index
+  auto index = (out_tv->memoryType() == MemoryType::Global)
+      ? Index::getConsumerIndex(in_tv->fuserTv(), loop_structure)
+      : Index::getProducerIndex(
+            in_tv->fuserTv(), out_tv->fuserTv(), loop_structure);
+
+  //  Get full extent for merged root domains
+  auto extent = getVectorizeExtent(index);
+
+  auto vector_size =
+      vec_tv->domain()->domain().back()->extent()->as<kir::Int>();
+
+  // Disable vectorize flag in child For-Loop
+  loop_structure.back()->setVectorize(false);
+
+  auto base_address_val = generateBaseIndex(index);
+  auto base_address =
+      setupNamedScalar(pred_ite->thenBody(), base_address_val, "base_address");
+
+  auto a = ir_builder.ceilDivExpr(base_address, vector_size);
+  auto b = ir_builder.mulExpr(a, vector_size);
+  auto shift_val = ir_builder.subExpr(b, base_address);
+  auto shift = setupNamedScalar(pred_ite->thenBody(), shift_val, "shift");
+
+  auto remaining_extent = ir_builder.subExpr(extent, shift);
+  auto remainder_val = ir_builder.modExpr(remaining_extent, vector_size);
+  auto remainder =
+      setupNamedScalar(pred_ite->thenBody(), remainder_val, "remainder");
+
+  auto last_index = ir_builder.subExpr(extent, vector_size);
+  auto threshold_val = ir_builder.subExpr(last_index, shift);
+  auto threshold =
+      setupNamedScalar(pred_ite->thenBody(), threshold_val, "threshold");
+
+  auto last_root_dim_index = setupNamedScalar(
+      pred_ite->thenBody(), index->indices().back(), "last_root_dim_index");
+
+  // Part A - Vectorize
```

Comment:
Is there any reason to do the vectorized part first? I'd imagine B->A->C would be more cache friendly in general, but maybe there's some other things to consider for GPUs.

Reply:
Not really. Christian like having the main part first, then the clean up sections.

Reply:
@csarofeen Is that preferred on GPUs than just linearly accessing?

Reply:
I thought it was a stylistic choice.

Reply:
OK, now I understand this is really just stylistic. I thought this ordering would mean we would first do the vectorized part and then the pre and post parts, which isn't actually the case, but they are just placed there in that order.

Reply:
Unclear if there would be any perf impacts. Vectorized reads is like concern #1 for memory access, so we're already >90% of the way there.

More complex uses there could be a couple effects, yes memory cache is one, instruction cache could be another. These might be at odds for which order we do these sections in.

Stylistically it's nice to see the vectorized section first in the generated code, and the cleanup after. Which order they're executed in is another conversation and hopefully the implementation doesn't firmly lock us into one choice.